Add ytick label right/left properties in matplotlibrc

esvhd opened the pull request:
## PR Summary

Add y-axis tick label positions into `matplotlibrc` for issue #10267 

New properties named as:
ytick.labelleft
ytick.labelright 

Will also work on a PR for x-axis tick label top/bottom properties for `matplotlibrc`

jklymak commented on lib/matplotlib/axes/_base.py:
```diff
@@ -559,13 +559,21 @@ def __init__(self, fig, rect,
             left=rcParams['ytick.left'] and rcParams['ytick.major.left'],
             right=rcParams['ytick.right'] and rcParams['ytick.major.right'],
+            labelleft=rcParams['ytick.labelleft'] and
+            rcParams['ytick.major.left'],
```
You should indent a 8 spaces or so here...

esvhd replied:
yep sounds good.

esvhd replied:
Actually when I indented like the below I got some PEP 8 complains. Was this what you meant? 

Alternatively I can have this in a single line but it breaches 80 line margin.

```
        labelleft=rcParams['ytick.labelleft'] and
            rcParams['ytick.major.left'],
```

jklymak replied:
Put the RHS in parentheses and then it should work with indenting...

jklymak commented:
This at least needs a test.  w/o looking at the rest of the code, I don't quite understand how this propagates to the axis drawing.

esvhd commented:
Essentially I traced down how `matplotlibrc` was read and how the properties are then set in `axes._base__init__()`.  So I made the change to pick up these two new properties. This did the trick. 

All up for testing, any suggestion on how this should be tested? Any example tests you can point me to perhaps?

jklymak commented:
Tests are in `matplotlib/lib/tests`.  If possible it'd be great if you could set up a testing environment on your local machine before pushing.    https://matplotlib.org/devel/testing.html

esvhd commented:
Good idea. 

Sent from my iPhone

> On 26 Jan 2018, at 21:45, Jody Klymak <notifications@github.com> wrote:
> 
> @jklymak commented on this pull request.
> 
> In lib/matplotlib/axes/_base.py:
> 
> >              which='minor')
>  
>          self.tick_params(
>              top=rcParams['xtick.top'] and rcParams['xtick.major.top'],
>              bottom=rcParams['xtick.bottom'] and rcParams['xtick.major.bottom'],
>              left=rcParams['ytick.left'] and rcParams['ytick.major.left'],
>              right=rcParams['ytick.right'] and rcParams['ytick.major.right'],
> +            labelleft=rcParams['ytick.labelleft'] and
> +            rcParams['ytick.major.left'],
> Put the RHS in parentheses and then it should work with indenting...
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

esvhd commented:
Added a test case, let me know what you think. 

I tested the test case **manually** in a notebook and all asserts worked. However, when I tried to run `pytest` I got the following error message:

```
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/anaconda3/lib/python3.6/site-packages/_pytest/main.py", line 101, in wrap_session
INTERNALERROR>     config._do_configure()
INTERNALERROR>   File "/usr/anaconda3/lib/python3.6/site-packages/_pytest/config.py", line 921, in _do_configure
INTERNALERROR>     self.hook.pytest_configure.call_historic(kwargs=dict(config=self))
INTERNALERROR>   File "/usr/anaconda3/lib/python3.6/site-packages/pluggy/__init__.py", line 628, in call_historic
INTERNALERROR>     res = self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/usr/anaconda3/lib/python3.6/site-packages/pluggy/__init__.py", line 222, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/anaconda3/lib/python3.6/site-packages/pluggy/__init__.py", line 216, in <lambda>
INTERNALERROR>     firstresult=hook.spec_opts.get('firstresult'),
INTERNALERROR>   File "/usr/anaconda3/lib/python3.6/site-packages/pluggy/callers.py", line 201, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/anaconda3/lib/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/anaconda3/lib/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/esvhd/git/esvhd_matplotlib/lib/matplotlib/testing/conftest.py", line 12, in pytest_configure
INTERNALERROR>     matplotlib._init_tests()
INTERNALERROR>   File "/home/esvhd/git/esvhd_matplotlib/lib/matplotlib/__init__.py", line 1448, in _init_tests
INTERNALERROR>     from matplotlib import ft2font
INTERNALERROR> ImportError: cannot import name 'ft2font'
```
When I ran `from matplotlib import ft2font` in a python console it actually works fine... Couldn't find much useful info googling this error... Any suggests? 

Thx

jklymak commented:
Are you sure you are using the same environment in each case?

esvhd commented:
Sorry not an expert in this clearly. Could you please clarify what you mean
by same environment? Thx

On Sun, Jan 28, 2018 at 9:37 PM, Jody Klymak <notifications@github.com>
wrote:

> Are you sure you are using the same environment in each case?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/10306#issuecomment-361098307>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG_aUd41pYaazH1w3EPMjqA_e85qXnkBks5tPOikgaJpZM4RqZFS>
> .
>

jklymak commented:
If you are using conda environments, then `fontconfig` has to be installed in your test environment.  If you aren't using environments you should :wink: https://conda.io/docs/user-guide/tasks/manage-environments.html

esvhd commented:
I see, conda environments. Yes, the env that has this code change has
`fontconfig`. I was being lazy so I hacked my installed `matplotlib` code
with these changes and tested in my notebooks manually (including running
the test code)..

On Sun, Jan 28, 2018 at 10:14 PM, Jody Klymak <notifications@github.com>
wrote:

> If you are using conda environments, then fontconfig has to be installed
> in your test environment. If you aren't using environments you should 😉
> https://conda.io/docs/user-guide/tasks/manage-environments.html
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/10306#issuecomment-361101055>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG_aUVD2OkcAdYySv-cz7q68HDvYQoPVks5tPPFCgaJpZM4RqZFS>
> .
>

jklymak commented:
OK, this is almost done.  This could use a "whats-new" entry: see `doc/users/next_whats_new`

Ideally, you would also have an example for it, particulaly one that differentiates it from `ax.yaxis.tick_right()`...  `examples/ticks_and_spines` I expect.  I note that we don't have any examples with `tick_right()`, so if you could include that too, it'd be fabulous.  

Thanks for all your work on this!

esvhd commented:
Sounds good, let me look into these. Thx for the pointers.

On Mon, Jan 29, 2018 at 5:02 PM, Jody Klymak <notifications@github.com>
wrote:

> OK, this is almost done. This could use a "whats-new" entry: see
> doc/users/next_whats_new
>
> Ideally, you would also have an example for it, particulaly one that
> differentiates it from ax.yaxis.tick_right()... examples/ticks_and_spines
> I expect. I note that we don't have any examples with tick_right(), so if
> you could include that too, it'd be fabulous.
>
> Thanks for all your work on this!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/10306#issuecomment-361313176>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG_aUYSj1vWQeXWzvK1Al8IrIp8iASfJks5tPfm-gaJpZM4RqZFS>
> .
>

esvhd commented:
Added example to show how to set these new properties in `rcParams` as well as using `tick_right()` or `tick_left()`. Created entry for what's new as well. Let me know how this looks. Thanks.

jklymak commented:
Hmmm.  Is this how the example is supposed to look?  I'd have expected the ticks and labels to be on the right for the last subplot (`ax[2]`).

https://5811-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/gallery/ticks_and_spines/tick_label_right.html#sphx-glr-gallery-ticks-and-spines-tick-label-right-py

QuLogic commented:
Not `Figure`, but `Axes` instantiation (though that's essentially the same time with `subplots`).

esvhd commented:
Fixed the example. Needed to set `rcParams` before the figure is created. Let me know how it looks.

jklymak commented on examples/ticks_and_spines/tick_label_right.py:
```diff
@@ -0,0 +1,33 @@
+============================================
+
+Example to show how to change two new properties in rcParams to put yaxis tick
+lables on either left or right by default.
```
I think the text should not refer to "new" properties (unless you plan to edit this in a year).  Also some formatting changes. (Note I didn't check the comment below for length).

Otherwise looks great!

```python
"""
We can use :rc:`ytick.labelright` (default False) and :rc:`ytick.right` (default False) 
and :rc:`ytick.labelleft` (default True) and :rc:`ytick.left` (default False) to 
control where on the axes ticks and their labels appear.  These properties can also be 
set in the ``.matplotlib/matplotlibrc``.

"""
```

esvhd commented:
Sure, will add these changes later today. 

Sent from my iPhone

> On 30 Jan 2018, at 23:37, Jody Klymak <notifications@github.com> wrote:
> 
> @jklymak requested changes on this pull request.
> 
> In examples/ticks_and_spines/tick_label_right.py:
> 
> > @@ -0,0 +1,33 @@
> +"""
> +============================================
> +Set default y-axis tick labels on the right
> +============================================
> +
> +Example to show how to change two new properties in rcParams to put yaxis tick
> +lables on either left or right by default.
> I think the text should not refer to "new" properties (unless you plan to edit this in a year). Also some formatting changes. (Note I didn't check the comment below for length).
> 
> Otherwise looks great!
> 
> """
> We can use :rc:`ytick.labelright` (default False) and :rc:`ytick.right` (default False) 
> and :rc:`ytick.labelleft` (default True) and :rc:`ytick.left` (default False) to 
> control where on the axes ticks and their labels appear.  These properties can also be 
> set in the ``.matplotlib/matplotlibrc``.
> 
> """
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

esvhd commented:
Hope this looks good. Thanks guys!

esvhd commented:
Great, thanks!

On Wed, Jan 31, 2018 at 10:42 PM, Jody Klymak <notifications@github.com>
wrote:

> *@jklymak* approved this pull request.
>
> This looks good to me.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/10306#pullrequestreview-93111951>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG_aUecXlUI2hpU4fIXsR6TrakQA06H_ks5tQOxygaJpZM4RqZFS>
> .
>

afvincent commented:
@esvhd Would it be possible to add to this PR the equivalent feature for the x-ticks?

Ideally, one could target the imminent Matplotlib 2.2rc1 if you can replicate what you have already done for the y-ticks quickly enough :). As time may be short, I would suggest to focus first on the feature: IMHO, extending the example may be done later if there were wishes to do so.

In a nutshell, it would be great to:

  1. Add `xtick.bottom` and `xtick.top` rcParams.
  2. Update the “What's new” entry.
  3. Update the `matplotlibrc.template` file as well.
  4. Add a test similar to `test_axes_tick_params_ylabelside`.
  5. *If time allows it*, possibly update the example:
    - either by explicitly demonstrating the new rcParams for x-ticks;
    - or simply by mentioning that there exist a similar feature for x-ticks (to the one for y-ticks).

afvincent commented:
Milestoning this PR for 2.2 provided that one can merge similar features for x-ticks **and** y-ticks.

esvhd commented:
I can take a look at this later this weekend. I think it’s pretty similar to ytick labels, but hope it won’t mess up the spacing between xlabel and title. Will put in an example also. Thx

Sent from my iPhone

> On 2 Feb 2018, at 03:23, Adrien F. Vincent <notifications@github.com> wrote:
> 
> Milestoning this PR for 2.2 provided that one can merge similar features for x-ticks and y-ticks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

afvincent commented:
About the issue with having the x-ticks and x-tick labels at the top that do not play well with a possible title, I do not think that adding new rcParams is very different from the current situation when the x-ticks and their label are tuned through the proper `Axes` methods. 

I guess that a quick workaround if some core devs are afraid of this issue might be to simply advertise that one can increase the title padding. A genuine first thought that comes to me is something like.
```python
import matplotlib.pyplot as plt

# If a fixed title padding is good enough
#plt.rcParams["axes.titlepad"] = 36  # 24 without x-label

fig, axs = plt.subplots(ncols=2)
for ax, xlbl in zip(axs.flat, ("X-label", "")):
    ax.plot([0, 1])

    ax.set_title("A title")
    ax.set_xlabel(xlbl)
    ax.xaxis.tick_top()
    ax.xaxis.set_label_position('top')

# Dynamic title padding if one prefers
for ax in axs.flat:
    ax.set_title(ax.get_title(), pad=36 if len(ax.get_xlabel()) > 0 else 24)

fig.tight_layout()
plt.show()
```

jklymak commented:
Or, you know, https://github.com/matplotlib/matplotlib/pull/9498 which will do this automatically.

afvincent commented:
I knew that it felt like something that was discussed recently 🐑! Thanks @jklymak for x-referencing.

esvhd commented:
Here we go. Added `xtick.labeltop` and `xtick.labelbottom` for `matplotlibrc`. Also added tests, examples and modified what's new. 

Currently this quick solution does not take care of title position. I understood that this is being worked on separately as @jklymak suggested. 

Let me know how this looks. Thanks everyone.

efiring commented:
I have a concern about the git history; someone with more git knowledge can comment on whether something should be done about it, and if so, what.

Normally we do not merge master into feature branches.  Instead, to keep them up to date, we rebase them on top of the master branch.

esvhd commented:
Would be great if someone can shed some light on this, not a git pro here. Thx

jklymak commented:
I think these are just merge commits and are OK.  OTOH these commits would be better if they were rebased into one commit.  

Oddly, given how much we ask people to squash things:

https://matplotlib.org/devdocs/devel/gitwash/development_workflow.html

doesn't describe the process, but this is what I do:

https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

QuLogic commented on examples/ticks_and_spines/tick_label_right.py:
```diff
@@ -0,0 +1,31 @@
+plt.rcParams['ytick.left'], plt.rcParams['ytick.labelleft'] = False, False
+
+
+x = np.array([x for x in range(10)])
```
`np.arange(10)`

QuLogic commented on examples/ticks_and_spines/tick_xlabel_top.py:
```diff
@@ -0,0 +1,26 @@
+plt.rcParams['xtick.bottom'], plt.rcParams['xtick.labelbottom'] = False, False
+plt.rcParams['xtick.top'], plt.rcParams['xtick.labeltop'] = True, True
+
+x = np.array([x for x in range(10)])
```
`np.arange(10)`

QuLogic commented on examples/ticks_and_spines/tick_label_right.py:
```diff
@@ -0,0 +1,31 @@
+import matplotlib.pyplot as plt
+import numpy as np
+
+plt.rcParams['ytick.right'], plt.rcParams['ytick.labelright'] = True, True
```
Chained assignment would be a bit simpler:
```
plt.rcParams['ytick.right'] = plt.rcParams['ytick.labelright'] = True 
```

QuLogic commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -5224,6 +5224,36 @@ def test_axes_tick_params_gridlines():
+    ax.tick_params(labelleft=False, labelright=True,
+                   which='minor')
+    # expects left false, right true
+    assert(ax.yaxis.majorTicks[0].label1On is False)
```
Unnecessary parentheses.

QuLogic commented on examples/ticks_and_spines/tick_label_right.py:
```diff
@@ -14,11 +14,11 @@
 
 
-x = np.array([x for x in range(10)])
+x = np.array([x for x in np.range(10)])
```
No, I mean you can replace the entire line with `np.arange(10)`.

esvhd commented:
Added suggested changes from @QuLogic

esvhd commented:
good idea. let me do that.

On Sun, Feb 4, 2018 at 9:03 PM, Elliott Sales de Andrade <
notifications@github.com> wrote:

> *@QuLogic* commented on this pull request.
> ------------------------------
>
> In examples/ticks_and_spines/tick_label_right.py
> <https://github.com/matplotlib/matplotlib/pull/10306#discussion_r165857775>
> :
>
> >
>
> -x = np.array([x for x in range(10)])
> +x = np.array([x for x in np.range(10)])
>
> No, I mean you can replace the entire line with np.arange(10).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/10306#pullrequestreview-93857932>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG_aUTt6HtkuzMQ3tI1xkBVs1nB3M4l4ks5tRhsxgaJpZM4RqZFS>
> .
>

esvhd commented:
Hi guys, checking in on this, is there any more work need on this PR, particularly from my side?
Thx

jklymak commented:
Excellent.  Thanks so much @esvhd!  Great contribution....

esvhd commented:
Great, thank you everyone. Glad to help! 

Sent from my iPhone

> On 7 Feb 2018, at 23:27, Jody Klymak <notifications@github.com> wrote:
> 
> Excellent. Thanks so much @esvhd! Great contribution....
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
